File: IKSolverExample.js
/**
 * IKSolverExample.js - Example usage of the IK Solver for skeleton manipulation
 * 
 * This demonstrates how to use the IK solver to move hands, arms, and fingers
 * by supplying current and target 3D positions
 */

// Import the IK Solver
const { IKSolver, Vector3D, Joint } = require('./IKSolver.js');

console.log("=== IK Solver Examples ===\n");

// Example 1: Simple Single Joint Movement
console.log("Example 1: Moving a single joint");
console.log("---------------------------------");
const wrist = new Joint(new Vector3D(0, -2, 0));
const currentPosition = wrist.position;
const targetPosition = new Vector3D(1, -1.5, 0.5);

console.log(`Current wrist position: (${currentPosition.x}, ${currentPosition.y}, ${currentPosition.z})`);
console.log(`Target position: (${targetPosition.x}, ${targetPosition.y}, ${targetPosition.z})`);

const newWrist = IKSolver.solveSingleJoint(wrist, targetPosition);
console.log(`New wrist position: (${newWrist.position.x}, ${newWrist.position.y}, ${newWrist.position.z})`);
console.log();

// Example 2: Arm IK using CCD Algorithm
console.log("Example 2: Moving an arm to reach a target (CCD Algorithm)");
console.log("-----------------------------------------------------------");
const shoulderPos = new Vector3D(0, 0, 0);
const armChain = IKSolver.createArmChain(shoulderPos, 1.0, 0.9);

console.log("Initial arm chain:");
armChain.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const targetHandPosition = new Vector3D(0.8, -1.5, 0.3);
console.log(`\nTarget hand position: (${targetHandPosition.x}, ${targetHandPosition.y}, ${targetHandPosition.z})`);

const solvedArmCCD = IKSolver.solveCCD(armChain, targetHandPosition, 20, 0.01);

console.log("\nArm chain after CCD IK:");
solvedArmCCD.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const endEffectorDistance = solvedArmCCD[solvedArmCCD.length - 1].position.distance(targetHandPosition);
console.log(`\nDistance to target: ${endEffectorDistance.toFixed(4)}`);
console.log();

// Example 3: Arm IK using FABRIK Algorithm
console.log("Example 3: Moving an arm to reach a target (FABRIK Algorithm)");
console.log("--------------------------------------------------------------");
const armChain2 = IKSolver.createArmChain(shoulderPos, 1.0, 0.9);

console.log("Initial arm chain:");
armChain2.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const targetHandPosition2 = new Vector3D(1.2, -1.0, -0.5);
console.log(`\nTarget hand position: (${targetHandPosition2.x}, ${targetHandPosition2.y}, ${targetHandPosition2.z})`);

const solvedArmFABRIK = IKSolver.solveFABRIK(armChain2, targetHandPosition2, 20, 0.01);

console.log("\nArm chain after FABRIK IK:");
solvedArmFABRIK.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const endEffectorDistance2 = solvedArmFABRIK[solvedArmFABRIK.length - 1].position.distance(targetHandPosition2);
console.log(`\nDistance to target: ${endEffectorDistance2.toFixed(4)}`);
console.log();

// Example 4: Finger IK
console.log("Example 4: Moving a finger to a target position");
console.log("------------------------------------------------");
const fingerBase = new Vector3D(0, 0, 0);
const fingerDirection = new Vector3D(1, 0, 0);
const fingerBoneLengths = [0.4, 0.3, 0.25]; // Proximal, middle, distal phalanges

const fingerChain = IKSolver.createFingerChain(fingerBase, fingerBoneLengths, fingerDirection);

console.log("Initial finger chain:");
fingerChain.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const targetFingerTip = new Vector3D(0.7, 0.5, 0.2);
console.log(`\nTarget finger tip position: (${targetFingerTip.x}, ${targetFingerTip.y}, ${targetFingerTip.z})`);

const solvedFinger = IKSolver.solveFABRIK(fingerChain, targetFingerTip, 15, 0.01);

console.log("\nFinger chain after IK:");
solvedFinger.forEach((joint, i) => {
    console.log(`  Joint ${i}: (${joint.position.x.toFixed(2)}, ${joint.position.y.toFixed(2)}, ${joint.position.z.toFixed(2)})`);
});

const fingerDistance = solvedFinger[solvedFinger.length - 1].position.distance(targetFingerTip);
console.log(`\nDistance to target: ${fingerDistance.toFixed(4)}`);
console.log();

// Example 5: Calculate Joint Angles
console.log("Example 5: Calculating joint angles");
console.log("------------------------------------");
const angles = IKSolver.calculateJointAngles(solvedArmCCD);
console.log("Joint angles in the solved arm chain:");
angles.forEach((angle, i) => {
    console.log(`  Joint ${i + 1} angle: ${angle.toFixed(2)}°`);
});
console.log();

// Example 6: Multiple Fingers (Hand)
console.log("Example 6: Moving multiple fingers on a hand");
console.log("---------------------------------------------");
const handBase = new Vector3D(0, 0, 0);
const fingers = [];
const fingerNames = ['Thumb', 'Index', 'Middle', 'Ring', 'Pinky'];
const fingerOffsets = [
    new Vector3D(-0.3, 0, 0.2),
    new Vector3D(-0.15, 0, 0),
    new Vector3D(0, 0, 0),
    new Vector3D(0.15, 0, 0),
    new Vector3D(0.3, 0, -0.1)
];

// Create finger chains
for (let i = 0; i < 5; i++) {
    const basePos = handBase.add(fingerOffsets[i]);
    const chain = IKSolver.createFingerChain(basePos, [0.4, 0.3, 0.25], new Vector3D(1, 0, 0));
    fingers.push({ name: fingerNames[i], chain: chain });
}

// Define target positions for each finger (e.g., making a fist)
const fingerTargets = [
    new Vector3D(0.3, 0.5, 0.3),   // Thumb
    new Vector3D(0.5, 0.4, 0.1),   // Index
    new Vector3D(0.6, 0.3, 0.0),   // Middle
    new Vector3D(0.5, 0.3, -0.1),  // Ring
    new Vector3D(0.4, 0.2, -0.2)   // Pinky
];

console.log("Solving IK for all fingers:");
fingers.forEach((finger, i) => {
    const solved = IKSolver.solveFABRIK(finger.chain, fingerTargets[i], 15, 0.01);
    const tipPos = solved[solved.length - 1].position;
    const distance = tipPos.distance(fingerTargets[i]);
    console.log(`  ${finger.name}: Target distance = ${distance.toFixed(4)}`);
});
console.log();

console.log("=== Examples Complete ===");

File: IKSolverExample.pas
program IKSolverExample;

{$APPTYPE CONSOLE}

{*******************************************************************************
  IKSolverExample - Example usage of the IK Solver for skeleton manipulation
  
  This demonstrates how to use the IK solver to move hands, arms, and fingers
  by supplying current and target 3D positions
*******************************************************************************}

uses
  System.SysUtils,
  IKSolver in 'IKSolver.pas';

procedure Example1_SimpleSingleJoint;
var
  Wrist, NewWrist: TJoint;
  CurrentPosition, TargetPosition: TVector3D;
begin
  WriteLn('Example 1: Moving a single joint');
  WriteLn('---------------------------------');
  
  CurrentPosition := TVector3D.Create(0, -2, 0);
  Wrist := TJoint.Create(CurrentPosition);
  TargetPosition := TVector3D.Create(1, -1.5, 0.5);
  
  WriteLn(Format('Current wrist position: (%.2f, %.2f, %.2f)', 
    [CurrentPosition.X, CurrentPosition.Y, CurrentPosition.Z]));
  WriteLn(Format('Target position: (%.2f, %.2f, %.2f)', 
    [TargetPosition.X, TargetPosition.Y, TargetPosition.Z]));
  
  NewWrist := TIKSolver.SolveSingleJoint(Wrist, TargetPosition);
  WriteLn(Format('New wrist position: (%.2f, %.2f, %.2f)', 
    [NewWrist.Position.X, NewWrist.Position.Y, NewWrist.Position.Z]));
  WriteLn;
end;

procedure Example2_ArmIKWithCCD;
var
  ShoulderPos, TargetHandPosition: TVector3D;
  ArmChain, SolvedArmCCD: TJointArray;
  I: Integer;
  EndEffectorDistance: Double;
begin
  WriteLn('Example 2: Moving an arm to reach a target (CCD Algorithm)');
  WriteLn('-----------------------------------------------------------');
  
  ShoulderPos := TVector3D.Create(0, 0, 0);
  ArmChain := TIKSolver.CreateArmChain(ShoulderPos, 1.0, 0.9);
  
  WriteLn('Initial arm chain:');
  for I := 0 to High(ArmChain) do
    WriteLn(Format('  Joint %d: (%.2f, %.2f, %.2f)', 
      [I, ArmChain[I].Position.X, ArmChain[I].Position.Y, ArmChain[I].Position.Z]));
  
  TargetHandPosition := TVector3D.Create(0.8, -1.5, 0.3);
  WriteLn(Format(#13#10'Target hand position: (%.2f, %.2f, %.2f)', 
    [TargetHandPosition.X, TargetHandPosition.Y, TargetHandPosition.Z]));
  
  SolvedArmCCD := TIKSolver.SolveCCD(ArmChain, TargetHandPosition, 20, 0.01);
  
  WriteLn(#13#10'Arm chain after CCD IK:');
  for I := 0 to High(SolvedArmCCD) do
  begin
    WriteLn(Format('  Joint %d:', [I]));
    WriteLn(Format('    Position: (%.2f, %.2f, %.2f)', 
      [SolvedArmCCD[I].Position.X, SolvedArmCCD[I].Position.Y, SolvedArmCCD[I].Position.Z]));
    WriteLn(Format('    Rotation: (%.2f°, %.2f°, %.2f°)', 
      [SolvedArmCCD[I].Rotation.X, SolvedArmCCD[I].Rotation.Y, SolvedArmCCD[I].Rotation.Z]));
  end;
  
  EndEffectorDistance := SolvedArmCCD[High(SolvedArmCCD)].Position.Distance(TargetHandPosition);
  WriteLn(Format(#13#10'Distance to target: %.4f', [EndEffectorDistance]));
  WriteLn;
end;

procedure Example3_ArmIKWithFABRIK;
var
  ShoulderPos, TargetHandPosition: TVector3D;
  ArmChain, SolvedArmFABRIK: TJointArray;
  I: Integer;
  EndEffectorDistance: Double;
begin
  WriteLn('Example 3: Moving an arm to reach a target (FABRIK Algorithm)');
  WriteLn('--------------------------------------------------------------');
  
  ShoulderPos := TVector3D.Create(0, 0, 0);
  ArmChain := TIKSolver.CreateArmChain(ShoulderPos, 1.0, 0.9);
  
  WriteLn('Initial arm chain:');
  for I := 0 to High(ArmChain) do
    WriteLn(Format('  Joint %d: (%.2f, %.2f, %.2f)', 
      [I, ArmChain[I].Position.X, ArmChain[I].Position.Y, ArmChain[I].Position.Z]));
  
  TargetHandPosition := TVector3D.Create(1.2, -1.0, -0.5);
  WriteLn(Format(#13#10'Target hand position: (%.2f, %.2f, %.2f)', 
    [TargetHandPosition.X, TargetHandPosition.Y, TargetHandPosition.Z]));
  
  SolvedArmFABRIK := TIKSolver.SolveFABRIK(ArmChain, TargetHandPosition, 20, 0.01);
  
  WriteLn(#13#10'Arm chain after FABRIK IK:');
  for I := 0 to High(SolvedArmFABRIK) do
  begin
    WriteLn(Format('  Joint %d:', [I]));
    WriteLn(Format('    Position: (%.2f, %.2f, %.2f)', 
      [SolvedArmFABRIK[I].Position.X, SolvedArmFABRIK[I].Position.Y, SolvedArmFABRIK[I].Position.Z]));
    WriteLn(Format('    Rotation: (%.2f°, %.2f°, %.2f°)', 
      [SolvedArmFABRIK[I].Rotation.X, SolvedArmFABRIK[I].Rotation.Y, SolvedArmFABRIK[I].Rotation.Z]));
  end;
  
  EndEffectorDistance := SolvedArmFABRIK[High(SolvedArmFABRIK)].Position.Distance(TargetHandPosition);
  WriteLn(Format(#13#10'Distance to target: %.4f', [EndEffectorDistance]));
  WriteLn;
end;

procedure Example4_FingerIK;
var
  FingerBase, FingerDirection, TargetFingerTip: TVector3D;
  FingerBoneLengths: TArray<Double>;
  FingerChain, SolvedFinger: TJointArray;
  I: Integer;
  FingerDistance: Double;
begin
  WriteLn('Example 4: Moving a finger to a target position');
  WriteLn('------------------------------------------------');
  
  FingerBase := TVector3D.Create(0, 0, 0);
  FingerDirection := TVector3D.Create(1, 0, 0);
  SetLength(FingerBoneLengths, 3);
  FingerBoneLengths[0] := 0.4;
  FingerBoneLengths[1] := 0.3;
  FingerBoneLengths[2] := 0.25;
  
  FingerChain := TIKSolver.CreateFingerChain(FingerBase, FingerBoneLengths, FingerDirection);
  
  WriteLn('Initial finger chain:');
  for I := 0 to High(FingerChain) do
    WriteLn(Format('  Joint %d: (%.2f, %.2f, %.2f)', 
      [I, FingerChain[I].Position.X, FingerChain[I].Position.Y, FingerChain[I].Position.Z]));
  
  TargetFingerTip := TVector3D.Create(0.7, 0.5, 0.2);
  WriteLn(Format(#13#10'Target finger tip position: (%.2f, %.2f, %.2f)', 
    [TargetFingerTip.X, TargetFingerTip.Y, TargetFingerTip.Z]));
  
  SolvedFinger := TIKSolver.SolveFABRIK(FingerChain, TargetFingerTip, 15, 0.01);
  
  WriteLn(#13#10'Finger chain after IK:');
  for I := 0 to High(SolvedFinger) do
  begin
    WriteLn(Format('  Joint %d:', [I]));
    WriteLn(Format('    Position: (%.2f, %.2f, %.2f)', 
      [SolvedFinger[I].Position.X, SolvedFinger[I].Position.Y, SolvedFinger[I].Position.Z]));
    WriteLn(Format('    Rotation: (%.2f°, %.2f°, %.2f°)', 
      [SolvedFinger[I].Rotation.X, SolvedFinger[I].Rotation.Y, SolvedFinger[I].Rotation.Z]));
  end;
  
  FingerDistance := SolvedFinger[High(SolvedFinger)].Position.Distance(TargetFingerTip);
  WriteLn(Format(#13#10'Distance to target: %.4f', [FingerDistance]));
  WriteLn;
end;

procedure Example5_CalculateJointAngles;
var
  ShoulderPos, TargetPosition: TVector3D;
  ArmChain, SolvedArm: TJointArray;
  Angles: TArray<Double>;
  I: Integer;
begin
  WriteLn('Example 5: Calculating joint angles');
  WriteLn('------------------------------------');
  
  ShoulderPos := TVector3D.Create(0, 0, 0);
  ArmChain := TIKSolver.CreateArmChain(ShoulderPos, 1.0, 0.9);
  TargetPosition := TVector3D.Create(0.8, -1.5, 0.3);
  
  SolvedArm := TIKSolver.SolveCCD(ArmChain, TargetPosition, 20, 0.01);
  Angles := TIKSolver.CalculateJointAngles(SolvedArm);
  
  WriteLn('Joint angles in the solved arm chain:');
  for I := 0 to High(Angles) do
    WriteLn(Format('  Joint %d angle: %.2f°', [I + 1, Angles[I]]));
  WriteLn;
end;

procedure Example6_MultipleFingers;
var
  HandBase: TVector3D;
  FingerOffsets: array[0..4] of TVector3D;
  FingerTargets: array[0..4] of TVector3D;
  FingerNames: array[0..4] of string;
  FingerBoneLengths: TArray<Double>;
  FingerChain, SolvedFinger: TJointArray;
  I: Integer;
  BasePos, TipPos: TVector3D;
  Distance: Double;
begin
  WriteLn('Example 6: Moving multiple fingers on a hand');
  WriteLn('---------------------------------------------');
  
  HandBase := TVector3D.Create(0, 0, 0);
  
  // Define finger offsets
  FingerOffsets[0] := TVector3D.Create(-0.3, 0, 0.2);
  FingerOffsets[1] := TVector3D.Create(-0.15, 0, 0);
  FingerOffsets[2] := TVector3D.Create(0, 0, 0);
  FingerOffsets[3] := TVector3D.Create(0.15, 0, 0);
  FingerOffsets[4] := TVector3D.Create(0.3, 0, -0.1);
  
  // Define finger names
  FingerNames[0] := 'Thumb';
  FingerNames[1] := 'Index';
  FingerNames[2] := 'Middle';
  FingerNames[3] := 'Ring';
  FingerNames[4] := 'Pinky';
  
  // Define target positions (e.g., making a fist)
  FingerTargets[0] := TVector3D.Create(0.3, 0.5, 0.3);
  FingerTargets[1] := TVector3D.Create(0.5, 0.4, 0.1);
  FingerTargets[2] := TVector3D.Create(0.6, 0.3, 0.0);
  FingerTargets[3] := TVector3D.Create(0.5, 0.3, -0.1);
  FingerTargets[4] := TVector3D.Create(0.4, 0.2, -0.2);
  
  SetLength(FingerBoneLengths, 3);
  FingerBoneLengths[0] := 0.4;
  FingerBoneLengths[1] := 0.3;
  FingerBoneLengths[2] := 0.25;
  
  WriteLn('Solving IK for all fingers:');
  for I := 0 to 4 do
  begin
    BasePos := HandBase.Add(FingerOffsets[I]);
    FingerChain := TIKSolver.CreateFingerChain(BasePos, FingerBoneLengths, TVector3D.Create(1, 0, 0));
    SolvedFinger := TIKSolver.SolveFABRIK(FingerChain, FingerTargets[I], 15, 0.01);
    TipPos := SolvedFinger[High(SolvedFinger)].Position;
    Distance := TipPos.Distance(FingerTargets[I]);
    WriteLn(Format('  %s: Target distance = %.4f', [FingerNames[I], Distance]));
  end;
  WriteLn;
end;

begin
  try
    WriteLn('=== IK Solver Examples ===');
    WriteLn;
    
    Example1_SimpleSingleJoint;
    Example2_ArmIKWithCCD;
    Example3_ArmIKWithFABRIK;
    Example4_FingerIK;
    Example5_CalculateJointAngles;
    Example6_MultipleFingers;
    
    WriteLn('=== Examples Complete ===');
    WriteLn;
    WriteLn('Press Enter to exit...');
    ReadLn;
  except
    on E: Exception do
    begin
      WriteLn('Error: ', E.Message);
      ReadLn;
    end;
  end;
end.
